enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,340 @@
+Finally the status section of the `insightsoperator.operator.openshift.io` CR is updated. Note that there is a retry logic in case of failed upload and after some time (and number of attempts) the Insights operator is marked as degraded.
+
+#### Updates required to allow on-demand data gathering
+The plan is to refactor the existing periodic data gathering to run in a separate Pod/Job. The existing Insights operator pod will still exist, but it would periodically do the following:
+- check the `insightsdatagather.config.openshift.io` CR to read current user configuration
+- create a new `datagather.insights.openshift.io` CR (in the `openshift-insights` namespace) based on the values from the above
+- spawn a new `job` to gather the Insights data, upload it to `console.redhat.com` and check that the archive was processed in console.redhat.com (this requires a new API endpoint to be implemented in https://console.redhat.com/docs/api/insights-results-aggregator/v2).
+  The new API endpoint will be identified by the cluster ID and will require `insightsRequestID` parameter. The endpoint will be queried with some delay (to allow processing)  
+- status attribute of the `datagather.insights.openshift.io` is updated within the job execution
+- wait for the job completion and if the processing part was successful then download the last Insights analysis report
+- check that the Insights analysis report ID is equal to the gathering job request ID. If the IDs are equal then update the corresponding job status with `AnalysisAvailable=True` condition and continue with the following step below, otherwise update the corresponding job status with `AnalysisAvailable=False` condition and do nothing 
+- based on the status of the last `datagather.insights.openshift.io` CR update the `insightsoperator.operator.openshift.io` CR and also the clusteroperator conditions
+- prune the old job definitions and `datagather.insights.openshift.io` CRs
+  
+As mentioned above the operator will work with CRs created only in the `openshift-insights` namespace and will create jobs only in this namespace. The periodically created jobs will have the `periodic` and current timestamp prefix. To start on-demand data gathering, the user must create a new CR with a unique name.
+
+### Risks and Mitigations
+
+Risk: Inconsistent Insights analysis results. Assume following situation:
+- periodic gathering is running in the Insights operator (let's call it gather A)
+- user runs a new on demand data gathering (let's call it gather B) e.g 30 seconds after gather A
+- gather A and B finish at essentially the same time
+- the external processing service (in the console.redhat.com) provides only one recent Insights analysis result, and there is no guarantee which (of the two) it is.
```

Comment:
Given a known finishTime stored in the resource and the InsightsRequestID I think we actually have enough information to know
1. did my gather ever get processed (I saw that above)
2. which DataGather is the currently served analysis result for

Is this still a problem?

Reply:
Having read the latest iteration of this enhancement, I believe this is also resolved, I think that's explained in the mitigation note below right?

Reply:
Correct.